### PR TITLE
Add Debouncing to Get Devices Call

### LINKED ISSUE #2625: Avoid fetching the device list multiple times in quick succession
I instrumented a local Fenix build to log all network activity, then poked around in my account settings to see what would happen. I wasn't able to receive tab via push because I didn't have firebase configured in my local build, but here's the network requests involved in me receiving a sent tab via "sync now":

```
02-11 19:51:16.905 19047 19086 W GeckoViewFetchClient: GV FETCH GET https://api.accounts.firefox.com/v1/account/devices
02-11 19:51:17.545 19047 19086 W GeckoViewFetchClient: GV FETCH RESPONSE GET https://api.accounts.firefox.com/v1/account/devices 200
02-11 19:51:17.555 19047 19086 I FxaDeviceConstellation: Executing: polling for device events
02-11 19:51:17.555 19047 19086 W GeckoViewFetchClient: GV FETCH GET https://api.accounts.firefox.com/v1/account/device/commands?index=8310664
02-11 19:51:17.805 19047 19086 W GeckoViewFetchClient: GV FETCH RESPONSE GET https://api.accounts.firefox.com/v1/account/device/commands?index=8310664
200 02-11 19:51:17.805 19047 19086 W GeckoViewFetchClient: GV FETCH GET https://api.accounts.firefox.com/v1/account/devices
02-11 19:51:18.015 19047 19086 W GeckoViewFetchClient: GV FETCH RESPONSE GET https://api.accounts.firefox.com/v1/account/devices 200                       
```

Note that this requests the device list twice in quick succession - once when asked to explicitly by the app in its "sync now" logic, and again, and a second time as part of (I assume) the handling of the received tab.

We should consider some sort of caching/debouncing logic similar to what we have in Desktop, to avoid fetching the device-list multiple times in quick succession.

┆Issue is synchronized with this [Jira Bug](https://jira.mozilla.com/browse/SYNC-815)
┆Story Points: 3
┆Epic: FxA crate in desktop
┆Sprint: SYNC - end 2020-05-08

### PR DESCRIPTION
**Pull request by lougeniaC64**
Fixes #2625
- adds debouncing to `get_devices()`

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - `automation/all_tests.sh` runs to completion and produces no failures
  - Note: For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/master/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/fxa-client/src/device.rs [hunk 1]
```diff
@@ -12,20 +12,37 @@ use crate::{
...
 use std::collections::{HashMap, HashSet};
 
+// An devices response is considered fresh for `DEVICES_FRESHNESS_THRESHOLD` ms.
+const DEVICES_FRESHNESS_THRESHOLD: u64 = 60_000; // 1 minute
+
 impl FirefoxAccount {
     /// Fetches the list of devices from the current account including
     /// the current one.
-    pub fn get_devices(&self) -> Result<Vec<Device>> {
+    pub fn get_devices(&mut self) -> Result<Vec<Device>> {
```
Comment (vladikoff) [inline components/fxa-client/src/device.rs:26]: in profile we also have `ignore_cache`, would we need something like this here as well https://github.com/mozilla/application-services/blob/f3fe12419e64b2c77f40398c8e54ca4a30672873/components/fxa-client/src/profile.rs#L21

File: components/fxa-client/src/oauth/attached_clients.rs [hunk 2]
```diff
@@ -9,6 +9,7 @@ use crate::{error::*, util, CachedResponse, FirefoxAccount};
 const ATTACHED_CLIENTS_FRESHNESS_THRESHOLD: u64 = 60_000; // 1 minute
 
 impl FirefoxAccount {
+    /// Fetches the list of attached clients connected to the current account.
```
Comment (lougeniaC64) [inline components/fxa-client/src/oauth/attached_clients.rs:12]: FYI: This is an unrelated addition. Just adding it because I was in the neighborhood.

File: components/fxa-client/src/device.rs (hunk 1 above)
Comment (rfk) [reply components/fxa-client/src/device.rs]: Desktop does have a similar `ignore_cache` facility FWIW

Comment (lougeniaC64) [reply components/fxa-client/src/device.rs]: @rfk From reading the comments in the bug I wasn't sure if this should be omitted until explicitly called for. I've added it.

File: CHANGES_UNRELEASED.md [hunk 3]
```diff
@@ -3,3 +3,9 @@
 # Unreleased Changes
 
 [Full Changelog](https://github.com/mozilla/application-services/compare/v0.58.2...master)
+
+## FxA Client
+- iOS: Added optional boolean argument `forceRefresh` to ignore caching for `fetchDevices`.
+- Added an optional `ttl` parameter to `getAccessToken` to limit the lifetime of the token. ([#3066](https://github.com/mozilla/application-services/pull/3066))
```
Comment (rfk) [inline CHANGES_UNRELEASED.md:9]: There's some stuff here about `ttl` which doesn't seem quite right, possibly copy-pasta or a bag merge from a previous changelog?

File: components/fxa-client/ffi/src/lib.rs [hunk 4]
```diff
@@ -414,15 +414,22 @@ pub extern "C" fn fxa_set_device_name(
...
+///
 /// # Safety
 ///
 /// A destructor [fxa_bytebuffer_free] is provided for releasing the memory for this
 /// pointer type.
 #[no_mangle]
-pub extern "C" fn fxa_get_devices(handle: u64, error: &mut ExternError) -> ByteBuffer {
+pub extern "C" fn fxa_get_devices(
+    handle: u64,
+    ignore_cache: bool,
```
Comment (rfk) [inline components/fxa-client/ffi/src/lib.rs:427]: For reasons that I won't pretend to understand in detail, we're not supposed to use `bool` at the FFI layer:

  https://github.com/mozilla/application-services/blob/master/docs/howtos/when-to-use-what-in-the-ffi.md#primitives

That `fxa_get_profile` currently does use it, is probably an oversight. Could you please adjust both `fxa_get_devices` and `fxa_get_profile` to use a `u8` here rather than `bool`, following the [example of the places FFI](https://github.com/mozilla/application-services/blob/5b130877dc5f5f8e4ebd8f0c1f4727d21a6c4e33/components/places/ffi/src/lib.rs#L266)? (Unfortunately this means converting it to a `u8` inside the [android](https://github.com/mozilla/application-services/blob/5b130877dc5f5f8e4ebd8f0c1f4727d21a6c4e33/components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt#L318) and ios binding layers 😞)

File: components/fxa-client/ios/FxAClient/FxAccount.swift [hunk 5]
```diff
@@ -67,9 +67,9 @@ class FxAccount: RustFxAccount {
         }
     }
 
-    override func fetchDevices() throws -> [Device] {
+    override func fetchDevices(forceRefresh: Bool = false) throws -> [Device] {
         return try notifyAuthErrors {
-            try super.fetchDevices()
+            try super.fetchDevices(forceRefresh: forceRefresh)
```
Comment (rfk) [inline components/fxa-client/ios/FxAClient/FxAccount.swift:72]: It pains me that we changed the name of this arg in the swift API, but 👍 for keeping it consistent with `getProfile` 😄

File: components/fxa-client/src/device.rs [hunk 6]
```diff
@@ -654,4 +674,71 @@ mod tests {
...
+
+        assert!(res.is_ok());
+        assert!(fxa.devices_cache.is_some());
+
+        let cached_devices_res = fxa.devices_cache.unwrap();
+        assert!(!cached_devices_res.response.is_empty());
+        assert!(cached_devices_res.cached_at > 0);
+
+        let cached_devices = &cached_devices_res.response[0];
+        assert_eq!(cached_devices.clone().id, "device1".to_string());
```
Comment (rfk) [inline components/fxa-client/src/device.rs:717]: Would it be possible to call `get_devices(false)` a second time here, and assert that it *didn't* make another network request?

File: components/fxa-client/ios/FxAClient/FxAccount.swift (hunk 5 above)
Comment (lougeniaC64) [reply components/fxa-client/ios/FxAClient/FxAccount.swift:72]: Yeah, I'm curious why iOS uses the fetchDevices/forceRefresh naming while android uses getDevices/ignoreCache. I almost didn't catch the iOS changes because of the difference. A part of me wants to rename the iOS functions but I also want to avoid causing breaking changes and possibly disregarding a sound design decision.

File: components/fxa-client/ffi/src/lib.rs (hunk 4 above)
Comment (lougeniaC64) [reply components/fxa-client/ffi/src/lib.rs:427]: Thanks! I was somewhat unsure about the FFI layer changes I made so this comment and the link are very helpful.

File: CHANGES_UNRELEASED.md (hunk 3 above)
Comment (lougeniaC64) [reply CHANGES_UNRELEASED.md:9]: Copy pasta for sure. You saw nothing.

File: components/fxa-client/ios/FxAClient/FxAccount.swift (hunk 5 above)
Comment (vladikoff) [reply components/fxa-client/ios/FxAClient/FxAccount.swift:72]: if renaming is sound then, IMHO we should clean those up at some point and report it in breaking changes

File: CHANGES_UNRELEASED.md [hunk 7]
```diff
@@ -3,3 +3,7 @@
 # Unreleased Changes
 
 [Full Changelog](https://github.com/mozilla/application-services/compare/v0.58.2...master)
+
+## FxA Client
+- iOS: Added optional boolean argument `forceRefresh` to ignore caching for `fetchDevices`. ([#3066](https://github.com/mozilla/application-services/pull/3066))
+- Android: Added option boolean argument `ignoreCache` to ignore caching for `getDevices`. ([#3066](https://github.com/mozilla/application-services/pull/3066))
```
Comment (rfk) [inline CHANGES_UNRELEASED.md:9]: Oh wow it's even worse when you see the two different names next to each other in the changelog like this 😆 

I will defer to @eoger on the wisdom or safety of attempting to rename the iOS version for consistency.

File: components/fxa-client/ios/FxAClient/FxAccount.swift (hunk 5 above)
Comment (rfk) [reply components/fxa-client/ios/FxAClient/FxAccount.swift:72]: /cc @eoger

File: CHANGES_UNRELEASED.md (hunk 7 above)
Comment (eoger) [reply CHANGES_UNRELEASED.md:9]: FWIW `CHANGES_UNRELEASED.md` is empty at the moment and needs to be rebased :)
But yeah I support this breaking change, iOS only uses the `forceRefresh` argument [in 1 place](https://searchfox.org/mozilla-mobile/rev/fd32c1b030795f42f68da92c36e9bdef97b52d29/firefox-ios/RustFxA/FxAWebView.swift#163) so it won't be too bad.
